crates/turborepo-api-client/src/lib.rs

```diff
@@ -240,7 +255,162 @@ impl APIClient {
+        hash: &str,
+        token: &str,
+        team_id: &str,
+        team_slug: Option<&str>,
+        use_preflight: bool,
+    ) -> Result<Response> {
+        self.get_artifact(hash, token, team_id, team_slug, use_preflight, Method::GET)
+            .await
+    }
+
+    pub async fn artifact_exists(
+        &self,
+        hash: &str,
+        token: &str,
+        team_id: &str,
+        team_slug: Option<&str>,
+        use_preflight: bool,
+    ) -> Result<Response> {
+        self.get_artifact(hash, token, team_id, team_slug, use_preflight, Method::HEAD)
+            .await
+    }
+
+    async fn get_artifact(
+        &self,
+        hash: &str,
+        token: &str,
+        team_id: &str,
+        team_slug: Option<&str>,
+        use_preflight: bool,
+        method: Method,
+    ) -> Result<Response> {
+        let mut request_url = self.make_url(&format!("/v8/artifacts/{}", hash));
+        let mut allow_auth = true;
+
+        if use_preflight {
+            let preflight_response = self
+                .do_preflight(token, &request_url, "GET", "Authorization, User-Agent")
+                .await?;
+
+            allow_auth = preflight_response.allow_auth;
+            request_url = preflight_response.location.to_string();
+        };
+
+        let mut request_builder = self
+            .client
+            .request(method, request_url)
+            .header("User-Agent", self.user_agent.clone());
+
+        if allow_auth {
+            request_builder = request_builder.header("Authorization", format!("Bearer {}", token));
+        }
+
+        request_builder = Self::add_team_params(request_builder, team_id, team_slug);
+
+        let response = retry::make_retryable_request(request_builder)
+            .await?
+            .error_for_status()?;
+
+        Ok(response)
+    }
+
+    pub async fn do_preflight(
+        &self,
+        token: &str,
+        request_url: &str,
+        request_method: &str,
+        request_headers: &str,
+    ) -> Result<PreflightResponse> {
+        let request_builder = self
+            .client
+            .request(Method::OPTIONS, request_url)
+            .header("User-Agent", self.user_agent.clone())
+            .header("Access-Control-Request-Method", request_method)
+            .header("Access-Control-Request-Headers", request_headers)
+            .header("Authorization", format!("Bearer {}", token));
+
+        let response = retry::make_retryable_request(request_builder).await?;
+
+        let headers = response.headers();
+        let location = if let Some(location) = headers.get("Location") {
+            let location = location.to_str()?;
+            Url::parse(location)?
+        } else {
+            response.url().clone()
+        };
+
+        let allowed_headers = headers
+            .get("Access-Control-Allow-Headers")
+            .map_or("", |h| h.to_str().unwrap_or(""));
+
+        let allow_auth = allowed_headers.to_lowercase().contains("authorization");
```

Comment:
`x-authorization-foo` also matches this. Needs boundary checking.

Let's get a test for this too.

Reply:
This is the Go behavior, not sure if we should fix it in the port or after

Reply:
It's a bug, and it's a straightforward regex:

```
let authorization_regex = Regex::new(r"(?i)(?:^|,) *authorization *(?:,|$)").unwrap();

authorization_regex.is_match(allowed_header);
```

Exploded:
```
(?i) - case insensitive
(?: - non-capturing alternation
  ^ - start of string
  | - or
  , - comma reset
)
 * - optional whitespace
authorization - literal string
 * - optional whitespace
(?: - non-capturing alternation
  , - comma
  | - or
  $ - end of string
)

***

Could be stricter on the whitespace if we wanted.